Optimize vectorized sorting - reduce code size, improve speed for large heaps

PeterSolMS opened the pull request:
There are two optimizations in this PR:

- reduction of code size in the bitonic sorters: by limiting the amount of inlining in this code, we can reduce overall code size in coreclr.dll by about 180 kB.

- dynamic packing: during sorting, we can switch to 32-bit sorting as soon as the address range in a partition is less 32 GB. This will only have an impact on large heaps or machines with many processors, because we already have a similar, but static optimization where we use 32-bit sorting if the overall address range in the ephemeral region is less than 32 GB. So this additional optimization will give improvements if the overall address range is greater than 32 GB initially, but becomes less during the sort. In this case, we get about a 1.6x improvement in sorting speed.

cc: @damageboy

PeterSolMS committed:
Improved vectorized sort - smaller bitonic sorters, dynamic packing/unpacking.

PeterSolMS committed:
Merge branch 'master' into vxsort-opt

PeterSolMS committed:
Merge branch 'master' into vxsort-opt

Maoni0 commented on src/coreclr/src/gc/vxsort/machine_traits.avx2.h:
```diff
@@ -12,8 +12,9 @@
 #include <immintrin.h>
 //#include <stdexcept>
+//#include <limits>
 #include <assert.h>
```
any reason to keep these 2 lines?

PeterSolMS committed:
Address code review feedback:
- remove commented out #include lines
- remove unreferenced template specializations for float, double, uint32_t and uint64_t

Mitigated trap with our implementation of numeric_limits - make it so we get a compile error when an instantiation of numeric_limits is referenced that is not specialized.
